### PR TITLE
[165] Sign-Up Form Validation- Frontend

### PR DESCRIPTION
**Pull request by Onyinye101**
## Task: `[165] Sign-Up Form Validation`
- `Validated the sign-up form using Javascript`

### CODE REVIEW

File: submissions/onyinyeChukwuka/javascript/signup.js [hunk 1]
```diff
@@ -0,0 +1,113 @@
...
+
+// to trigger the required error messages when the user clicks the signup button
+signupbutton.addEventListener('click', () => {
+    let emailValid = validateEmail(email);
+    let numberValid = validateNumber(phoneNumber);
+    let passwordValid =  validatePassword(password);
+    let confirmPasswordValid = validateConfirmPassword(confirmPassword, password);
+    let firstNameValid = validateFirstName(firstName);
+    let lastNameValid = validateLastName(lastName);
+    if (emailValid && numberValid && passwordValid && confirmPasswordValid && firstNameValid && lastNameValid) {
```
Comment (adaugochi) [inline submissions/onyinyeChukwuka/javascript/signup.js:17]: This check should be for debugging purpose. There is no need for this to be in your code

File: submissions/onyinyeChukwuka/javascript/signup.js [hunk 2]
```diff
@@ -0,0 +1,113 @@
...
+let lastName = document.getElementById('lastname');
+let phoneNumber = document.getElementById('phonenumber');
+let email = document.getElementById('email');
+let password = document.getElementById('password');
+let confirmPassword = document.getElementById('cpassword');
+let signupbutton = document.getElementById('signupbutton');
+
+// to trigger the required error messages when the user clicks the signup button
+signupbutton.addEventListener('click', () => {
+    let emailValid = validateEmail(email);
```
Comment (adaugochi) [inline submissions/onyinyeChukwuka/javascript/signup.js:11]: Kindly remove all variable names

**Review by adaugochi — COMMENTED** [review]
Well done.

But, you don't need to write separate functions for each input field validate. This defeat the use of functions. Functions are means to be reusable.

File: submissions/onyinyeChukwuka/javascript/signup.js [hunk 3]
```diff
@@ -0,0 +1,113 @@
+let firstName = document.getElementById('firstname');
+let lastName = document.getElementById('lastname');
+let phoneNumber = document.getElementById('phonenumber');
+let email = document.getElementById('email');
+let password = document.getElementById('password');
+let confirmPassword = document.getElementById('cpassword');
+let signupbutton = document.getElementById('signupbutton');
+
+// to trigger the required error messages when the user clicks the signup button
+signupbutton.addEventListener('click', () => {
```
Comment (adaugochi) [inline submissions/onyinyeChukwuka/javascript/signup.js:10]: kindly listen for a `submit` event

**Review by adaugochi — CHANGES_REQUESTED** [review]

**Review by adaugochi — CHANGES_REQUESTED** [review]
Well done.

But, you don't need to write separate functions for each input field validate. Functions are meant to be reusable. Since almost all the function you have written do the same thing or are exactly alike, you can simply write a single function and pass in those things that make them differ a little as a parameter.

File: submissions/onyinyeChukwuka/javascript/signup.js [hunk 4]
```diff
@@ -0,0 +1,59 @@
...
+});
+
+
+// to print out individual error messages according to the specific text fields
+validateInput = (element, emptyElementMessage, validationFailedMessage, validationRegex) => {
+    if (element.value === "") {
+        element.nextElementSibling.innerHTML = emptyElementMessage;
+        return false;
+    }
+    if (element === phoneNumber || element === email || element === password) {
```
Comment (adaugochi) [inline submissions/onyinyeChukwuka/javascript/signup.js:30]: Well done. 

You don't need to do this checks i.e `element === phoneNumber || element === email || element === password`

File: submissions/onyinyeChukwuka/javascript/signup.js [hunk 5]
```diff
@@ -0,0 +1,59 @@
...
+        if (!validationRegex.test(element.value)) {
```
Comment (adaugochi) [inline submissions/onyinyeChukwuka/javascript/signup.js:31]: This line is where your code should start from

File: submissions/onyinyeChukwuka/javascript/signup.js [hunk 6]
```diff
@@ -0,0 +1,59 @@
...
+    } else if (element === confirmPassword) {
+        if (confirmPassword.value !== password.value) {
+            element.nextElementSibling.innerHTML = validationFailedMessage;
+            return false;
+        } else {
+            element.nextElementSibling.innerHTML = "";
+            return true;
+        }
+    }
+    else {
```
Comment (adaugochi) [inline submissions/onyinyeChukwuka/javascript/signup.js:47]: The code on the else block is irrelevant. the code on line 31 is already handling the validation of regex for all elements field whether firstname, lastname, email and so on

File: submissions/onyinyeChukwuka/javascript/signup.js [hunk 7]
```diff
@@ -0,0 +1,59 @@
...
+    }
+    if (element === phoneNumber || element === email || element === password) {
+        if (!validationRegex.test(element.value)) {
+            element.nextElementSibling.innerHTML = validationFailedMessage;
+            return false;
+        } else {
+            element.nextElementSibling.innerHTML = "";
+            return true;
+        }
+    } else if (element === confirmPassword) {
```
Comment (adaugochi) [inline submissions/onyinyeChukwuka/javascript/signup.js:38]: Confirm password should be done in a separate function

**Review by adaugochi — CHANGES_REQUESTED** [review]

File: submissions/onyinyeChukwuka/javascript/signup.js [hunk 8]
```diff
@@ -0,0 +1,48 @@
...
+const passwordRegex = (/^(?=.*\d)(?=.*[a-z])(?=.*[A-Z]).{6,}$/);
+
+// to trigger the required error messages when the user clicks the signup button
+signupbutton.addEventListener('submit', () => {
+    validateInput(email, "Please enter your Email Address", "Please enter a valid email", emailRegex);
+    validateInput(firstName, "Please enter your First Name", "Please enter a valid first name", nameRegex);
+    validateInput(lastName, "Please enter your Last Name", "Please enter a valid last name", nameRegex);
+    validateInput(phoneNumber, "Please enter your Phone Number", "Please enter a valid phone number", phoneRegex);
+    validateInput(password, "Please enter your Password", "Your password should have a minimum of 6 characters, 1 capital letter, 1 special character eg @ and 1 number", passwordRegex);
+    // validateInput(confirmPassword, "Please confirm your Password", "Your passwords don't match", passwordRegex);
```
Comment (ToluAkin) [inline submissions/onyinyeChukwuka/javascript/signup.js:20]: Remove this unused comment. @Onyinye101

**Review by ToluAkin — CHANGES_REQUESTED** [review]
Well done on the work done so far.

**Review by ToluAkin — APPROVED** [review]

**Pull request merged**